src/test/scala/DSATest.scala

```diff
@@ -0,0 +1,707 @@
+import analysis.{AddressRange, DSC, DSG, DSN, DataLocation, Derm, HeapLocation}
+import ir.Endian.BigEndian
+import ir.{Assign, BVADD, BinaryExpr, BitVecLiteral, CFGPosition, ConvertToSingleProcedureReturn, DirectCall, Memory, MemoryAssign, MemoryLoad, Register, SharedMemory}
+import org.scalatest.funsuite.AnyFunSuite
+import test_util.TestUtil
+import ir.dsl.*
+import specification.Specification
+import util.{BASILConfig, BoogieGeneratorConfig, ILLoadingConfig, IRContext, RunUtils, StaticAnalysisConfig}
+
+/**
+ * This is the test suite for testing DSA functionality
+ * The tests follow a general pattern of running BASIL analyses on a test program
+ * and then asserting properties about the Data Structure Graph (DSG) of the function produced at
+ * different levels
+ *
+ * DSA has three phases.
+ * BASILRESULT.analysis.get.local is the set of graphs from the end of the local phase
+ * BASILRESULT.analysis.get.bu is the set of graphs from the end of the bottom-up phase
+ * BASILRESULT.analysis.get.td is the set of graphs from the end of the top-down phase
+ *
+ */
+class DSATest extends AnyFunSuite, TestUtil {
+
+  // Local DSA tests
+  test("basic pointer") {
+    val results = RunUtils.loadAndTranslate(
+      BASILConfig(
+        loading = ILLoadingConfig(
+          inputFile = "examples/basicpointer/basicpointer.adt",
+          relfFile = "examples/basicpointer/basicpointer.relf",
+          specFile = None,
+          dumpIL = None,
+        ),
+        staticAnalysis = Some(StaticAnalysisConfig()),
+        boogieTranslation = BoogieGeneratorConfig(),
+        outputPrefix = "boogie_out",
+      )
+    )
+    val program = results.ir.program
+
+    // the dsg of the main procedure after the local phase
+    val dsg = results.analysis.get.locals.get(program.mainProcedure)
+
+
+
+    // dsg.formals(R29) is the slice representing formal R29
+    val R29formal = dsg.adjust(dsg.formals(R29))
+
+
+    // cells representing the stack at various offsets
+    val stack0 = dsg.find(dsg.stackMapping(0).cells(0)) // R31
+    val stack8 = dsg.find(dsg.stackMapping(8).cells(0)) //  R31 + 8
+    val stack40 = dsg.find(dsg.stackMapping(40).cells(0))//  R31 + 40
+    val stack32 = dsg.find(dsg.stackMapping(32).cells(0)) //  R31 + 32
+    val stack24 = dsg.find(dsg.stackMapping(24).cells(0)) //  R31 + 24 and Malloc
+
+    assert(dsg.adjust(stack0.getPointee).equals(R29formal)) // R31 points to the frame pointer
+    assert(dsg.adjust(stack8.getPointee).equals(dsg.adjust(dsg.formals(R30)))) // R31 + 8 points to the link register
+
+
+    assert(dsg.adjust(stack32.getPointee).equals(stack24))  // R31 + 32 points to R31 + 24, 00000457
+    assert(stack24.node.get.collapsed) // 00000497 collapses stack24 concatenation is currently unhandled, any objects referenced in an unhandled operation are collapsed
```

Comment:
Maybe it would be better to test the analysis on examples that can be handled properly?

Reply:
This is part of DSA's approach to handling more instructions while remaining sound.  it's the last instruction in figure 3.5 of the Lattner dissertation.

Reply:
https://github.com/UQ-PAC/BASIL/pull/222#discussion_r1751272452

Reply:
The concatenation is only because the input you are using was incorrectly lifted by a very old version of BAP. If you want to test this behaviour I suggest writing an example designed to test it.